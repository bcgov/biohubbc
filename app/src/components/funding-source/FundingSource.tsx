import Box from '@mui/material/Box';
import Grid from '@mui/material/Grid';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import Typography from '@mui/material/Typography';
import { DATE_FORMAT } from 'constants/dateTimeFormats';
import { getFormattedAmount, getFormattedDateRangeString } from 'utils/Utils';

export interface IFundingSource {
  id: number;
  agency_name?: string;
  investment_action_category_name?: string;
  funding_amount?: number;
  start_date?: string;
  end_date?: string;
  agency_project_id: string;
  first_nations_name?: string;
}

interface IFundingSourceProps {
  funding_sources: IFundingSource[];
}

/**
 * Funding source content for a project.
 *
 * @return {*}
 */
const FundingSource = (props: IFundingSourceProps) => {
  const hasFundingSources = props.funding_sources.length > 0;
  return (
    <>
      <List disablePadding>
        {hasFundingSources &&
          props.funding_sources.map((item: IFundingSource) => (
            <ListItem disableGutters divider key={item.id}>
              <Box flex="1 1 auto">
                <Box pb={1.25}>
                  <Typography component="span">
                    {item.agency_name ?? item.first_nations_name}
                    {item.investment_action_category_name &&
                      item.investment_action_category_name !== 'Not Applicable' && (
                        <Typography component="span">&nbsp;({item.investment_action_category_name})</Typography>
                      )}
                  </Typography>
                </Box>
                <Box component="dl" m={0}>
                  <Grid container spacing={1}>
                    <Grid item sm={6}>
                      <Typography component="dt" variant="subtitle2" color="textSecondary">
                        Project ID
                      </Typography>
                      <Typography component="dd">{item.agency_project_id || 'No Agency Project ID'}</Typography>
                    </Grid>
                    <Grid item sm={6}>
                      {item.start_date && item.end_date && (
                        <>
                          <Typography component="dt" variant="subtitle2" color="textSecondary">
                            Timeline
                          </Typography>
                          <Typography component="dd">
                            {getFormattedDateRangeString(
                              DATE_FORMAT.ShortMediumDateFormat,
                              item.start_date,
                              item.end_date
                            )}
                          </Typography>
                        </>
                      )}
                    </Grid>
                    <Grid item sm={12}>
                      {item.funding_amount && (
                        <>
                          <Typography component="dt" variant="subtitle2" color="textSecondary">
                            Funding Amount
                          </Typography>
                          <Typography component="dd">{getFormattedAmount(item.funding_amount)}</Typography>
                        </>
                      )}
                    </Grid>
                  </Grid>
                </Box>
              </Box>
            </ListItem>
          ))}

        {!hasFundingSources && (
          <ListItem disableGutters>
            <Typography>No Funding Sources</Typography>
          </ListItem>
        )}
      </List>
    </>
  );
};

export default FundingSource;
