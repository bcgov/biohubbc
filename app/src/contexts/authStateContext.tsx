import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import React from 'react';

export interface IAuthState {
  ready?: boolean;
}

export const AuthStateContext = React.createContext<IAuthState>({
  ready: false
});

export const AuthStateContextProvider: React.FC = (props) => {
  const keycloakWrapper = useKeycloakWrapper();

  return (
    <AuthStateContext.Provider value={{ ready: keycloakWrapper.keycloak?.authenticated }}>
      {props.children}
    </AuthStateContext.Provider>
  );
};
