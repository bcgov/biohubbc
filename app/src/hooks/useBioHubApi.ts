import { useKeycloak } from '@react-keycloak/web';
import axios from 'axios';
import { IProject, IProjectPostObject } from 'interfaces/project-interfaces';
import { IActivity, ICreateActivity, ICreateProjectResponse, ITemplate } from 'interfaces/useBioHubApi-interfaces';
import moment from 'moment';
import { useMemo } from 'react';

const API_HOST = process.env.REACT_APP_API_HOST;
const API_PORT = process.env.REACT_APP_API_PORT;

const API_URL = (API_PORT && `${API_HOST}:${API_PORT}`) || API_HOST || 'api-dev-biohubbc.apps.silver.devops.gov.bc.ca';

/**
 * Returns an instance of axios with baseURL and authorization headers set.
 *
 * @return {*}
 */
const useApi = () => {
  const { keycloak } = useKeycloak();
  const instance = useMemo(() => {
    return axios.create({
      headers: {
        Authorization: `Bearer ${keycloak?.token}`
      },
      baseURL: API_URL
    });
  }, [keycloak]);

  return instance;
};

/**
 * Returns a set of supported api methods.
 *
 * @return {object} object whose properties are supported api methods.
 */
export const useBiohubApi = () => {
  const api = useApi();

  /**
   * Get all projects.
   *
   * @return {*}  {Promise<IProject[]>}
   */
  const getProjects = async (): Promise<IProject[]> => {
    // const { data } = await api.get(`/api/projects`);

    // return data;

    // TODO: stub for development
    return [
      {
        id: 1,
        name: 'Project Name 1',
        objectives: 'Project Objectives 1',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 1',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 1',
        caveats: 'Caveats 1',
        comments: 'Comments 1'
      },
      {
        id: 2,
        name: 'Project Name 2',
        objectives: 'Project Objectives 2',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 2',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 2',
        caveats: 'Caveats 2',
        comments: 'Comments 2'
      },
      {
        id: 3,
        name: 'Project Name 3',
        objectives: 'Project Objectives 3',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 3',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 3',
        caveats: 'Caveats 3',
        comments: 'Comments 3'
      },
      {
        id: 4,
        name: 'Project Name 4',
        objectives: 'Project Objectives 4',
        scientific_collection_permit_number: '123456',
        management_recovery_action: 'A',
        location_description: 'Location Description 4',
        start_date: moment().toISOString(),
        end_date: moment().toISOString(),
        results: 'Results 4',
        caveats: 'Caveats 4',
        comments: 'Comments 4'
      }
    ];
  };

  /**
   * Get a project based on its ID.
   *
   * @param {projectId} projectId
   * @return {*}  {Promise<IProject>}
   */
  const getProject = async (projectId: number): Promise<IProject> => {
    const { data } = await api.get(`/api/project/${projectId}`);

    return data;
  };

  /**
   * Create a new project.
   *
   * @param {IProjectPostObject} project
   * @return {*}  {Promise<ICreateProjectResponse>}
   */
  const createProject = async (project: IProjectPostObject): Promise<ICreateProjectResponse> => {
    const { data } = await api.post('/api/project', project);

    return data;
  };

  /**
   * Get a template based on its ID.
   *
   * @param {templateId} templateId
   * @return {*}  {Promise<ITemplate>}
   */
  const getTemplate = async (templateId: number): Promise<ITemplate> => {
    const { data } = await api.get(`/api/template/${templateId}`);

    return data;
  };

  /**
   * Create a new activity record.
   *
   * @param {ICreateActivity} activity
   * @return {*}  {Promise<IActivity>}
   */
  const createActivity = async (activity: ICreateActivity): Promise<IActivity> => {
    const { data } = await api.post('/api/activity', activity);

    return data;
  };

  /**
   * Fetch all code sets.
   *
   * @return {*}  {Promise<any>}
   */
  const getAllCodes = async (): Promise<any> => {
    const { data } = await api.get('/api/codes/');

    return data;
  };

  /**
   * Fetch the api json-schema spec.
   *
   * @return {*}  {Promise<any>}
   */
  const getApiSpec = async (): Promise<any> => {
    const { data } = await api.get('/api/api-docs/');

    return data;
  };

  return {
    getProjects,
    getProject,
    getTemplate,
    createProject,
    createActivity,
    getAllCodes,
    getApiSpec
  };
};
