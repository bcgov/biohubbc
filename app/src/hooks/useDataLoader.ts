import { useState } from 'react';
import { AsyncFunction, useAsync } from './useAsync';
import useIsMounted from './useIsMounted';

export type DataLoader<AFArgs extends any[], AFResponse = unknown, AFError = unknown> = {
  /**
   * The response of the `fetchData` call.
   *
   * @type {(AFResponse | undefined)}
   */
  data: AFResponse | undefined;
  /**
   * The error caught if the `fetchData` call throws.
   *
   * @type {(AFError | unknown)}
   */
  error: AFError | unknown;
  /**
   * `true` if the `fetchData` function is currently executing.
   *
   * @type {boolean}
   */
  isLoading: boolean;
  /**
   * `true` if the `fetchData` function has finished executing.
   *
   * @type {boolean}
   */
  isReady: boolean;
  /**
   * Executes the `fetchData` function once, only if it has never been called before. Does nothing if called again.
   */
  load: (...args: AFArgs) => void;
  /**
   * Executes the `fetchData` function again.
   */
  refresh: (...args: AFArgs) => void;
  /**
   * Clears any errors caught from a failed `fetchData` call.
   */
  clear: () => void;
};

/**
 * Hook that wraps an async function.
 *
 * Note: Runs each time `refresh` is called.
 *
 * Note: This hook will prevent additional calls to `fetchData` if an existing call is in progress.
 *
 * @export
 * @template AFArgs `AsyncFunction` argument types.
 * @template AFResponse `AsyncFunction` response type.
 * @template AFError `AsyncFunction` error type.
 * @param {AsyncFunction<AFArgs, AFResponse>} fetchData An async function.
 * @param {((error: AFError | unknown) => void)} [onError] An optional error handler function that will be called if the
 * `fetchData` function throws an error.
 * - If set to `true`, the `fetchData` function will run on initial load, and each time `refresh` is called.
 * - If set to `false` the `fetchData` function will run each time `refresh` is called.
 * @return {*}  {DataLoader<AFArgs, AFResponse, AFError>}
 */
export default function useDataLoader<AFArgs extends any[], AFResponse = unknown, AFError = unknown>(
  fetchData: AsyncFunction<AFArgs, AFResponse>,
  onError?: (error: AFError | unknown) => void
): DataLoader<AFArgs, AFResponse, AFError> {
  const [data, setData] = useState<AFResponse>();
  const [error, setError] = useState<AFError | unknown>();
  const [isLoading, setIsLoading] = useState(false);
  const [isReady, setIsReady] = useState(false);
  const [isOneTimeLoad, setOneTimeLoad] = useState(false);

  const isMounted = useIsMounted();

  const getData = useAsync(fetchData);

  const loadData = async (...args: AFArgs) => {
    try {
      setIsLoading(true);

      const response = await getData(...args);

      if (!isMounted) {
        return;
      }

      setData(response);
    } catch (error) {
      if (!isMounted) {
        return;
      }

      setError(error);

      onError?.(error);
    } finally {
      setIsLoading(false);
      setIsReady(true);
    }
  };

  const load = (...args: AFArgs) => {
    if (isOneTimeLoad) {
      return;
    }

    setOneTimeLoad(true);
    loadData(...args);
  };

  const refresh = (...args: AFArgs) => {
    setError(undefined);
    setIsLoading(false);
    setIsReady(false);
    loadData(...args);
  };

  const clear = () => {
    setError(undefined);
  };

  return { data, error, isLoading, isReady, load, refresh, clear };
}
