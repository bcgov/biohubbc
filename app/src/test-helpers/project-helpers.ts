import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';

export const getProjectForViewResponse: IGetProjectForViewResponse = {
  id: 1,
  project: {
    project_name: 'Test Project Name',
    project_type: '1',
    start_date: '1998-10-10',
    end_date: '2021-02-26',
    project_activities: [1],
    completion_status: 'Active',
    publish_date: '2021-01-26'
  },
  permit: {
    permits: [
      {
        permit_number: '123',
        permit_type: 'Permit type'
      }
    ]
  },
  location: {
    location_description: 'Location description',
    geometry: []
  },
  objectives: {
    objectives: 'Et ad et in culpa si',
    caveats: 'sjwer bds'
  },
  coordinator: {
    first_name: 'Amanda',
    last_name: 'Christensen',
    email_address: 'amanda@christensen.com',
    coordinator_agency: 'Amanda and associates',
    share_contact_details: 'true'
  },
  iucn: {
    classificationDetails: [
      {
        classification: 'classification',
        subClassification1: 'sub classification 1',
        subClassification2: 'sub classification 2'
      },
      {
        classification: 'classification again',
        subClassification1: 'sub classification 1 again',
        subClassification2: 'sub classification 2 again'
      }
    ]
  },
  funding: {
    fundingSources: [
      {
        id: 0,
        agency_id: 1,
        agency_name: 'agency name',
        agency_project_id: 'ABC123',
        investment_action_category: 1,
        investment_action_category_name: 'investment action',
        funding_amount: 333,
        start_date: '2000-04-14',
        end_date: '2021-04-13',
        revision_count: 1
      }
    ]
  },
  partnerships: {
    indigenous_partnerships: ['partner 1', 'partner 2'],
    stakeholder_partnerships: ['partner 3', 'partner 4']
  }
};
