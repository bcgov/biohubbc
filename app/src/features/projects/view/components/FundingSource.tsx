import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { mdiPencilOutline, mdiPlus, mdiTrashCanOutline } from '@mdi/js';
import Icon from '@mdi/react';
import EditDialog from 'components/dialog/EditDialog';
import { IErrorDialogProps } from 'components/dialog/ErrorDialog';
import { IYesNoDialogProps } from 'components/dialog/YesNoDialog';
import { H3ButtonToolbar } from 'components/toolbar/ActionToolbars';
import { DATE_FORMAT } from 'constants/dateTimeFormats';
import { AddFundingI18N, DeleteProjectFundingI18N, EditFundingI18N } from 'constants/i18n';
import { DialogContext } from 'contexts/dialogContext';
import ProjectFundingItemForm, {
  IProjectFundingFormArrayItem,
  ProjectFundingFormArrayItemInitialValues,
  ProjectFundingFormArrayItemYupSchema
} from 'features/projects/components/ProjectFundingItemForm';
import { APIError } from 'hooks/api/useAxios';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useContext, useState } from 'react';
import { getFormattedAmount, getFormattedDate, getFormattedDateRangeString } from 'utils/Utils';

const useStyles = makeStyles((theme: Theme) => ({
  fundingSourceTable: {
    '& .MuiTableCell-root': {
      verticalAlign: 'middle'
    }
  }
}));

export interface IProjectFundingProps {
  projectForViewData: IGetProjectForViewResponse;
  codes: IGetAllCodeSetsResponse;
  refresh: () => void;
}

/**
 * Funding source content for a project.
 *
 * @return {*}
 */
const FundingSource: React.FC<IProjectFundingProps> = (props) => {
  const classes = useStyles();

  const {
    projectForViewData: { funding, id },
    codes
  } = props;

  const restorationTrackerApi = useRestorationTrackerApi();

  const dialogContext = useContext(DialogContext);

  const defaultErrorDialogProps = {
    dialogTitle: EditFundingI18N.editErrorTitle,
    dialogText: EditFundingI18N.editErrorText,
    open: false,
    onClose: () => {
      dialogContext.setErrorDialog({ open: false });
    },
    onOk: () => {
      dialogContext.setErrorDialog({ open: false });
    }
  };

  const showErrorDialog = (textDialogProps?: Partial<IErrorDialogProps>) => {
    dialogContext.setErrorDialog({ ...defaultErrorDialogProps, ...textDialogProps, open: true });
  };

  const defaultYesNoDialogProps = {
    dialogTitle: DeleteProjectFundingI18N.deleteTitle,
    dialogText: DeleteProjectFundingI18N.deleteText,
    open: false,
    onClose: () => dialogContext.setYesNoDialog({ open: false }),
    onNo: () => dialogContext.setYesNoDialog({ open: false }),
    onYes: () => handleDeleteDialogYes()
  };

  const showYesNoDialog = (yesNoDialogProps?: Partial<IYesNoDialogProps>) => {
    dialogContext.setYesNoDialog({ ...defaultYesNoDialogProps, ...yesNoDialogProps });
  };

  const [fundingFormData, setFundingFormData] = useState({
    index: 0,
    values: ProjectFundingFormArrayItemInitialValues
  });

  const [openEditDialog, setOpenEditDialog] = useState(false);

  const handleDialogEditOpen = async (itemIndex: number) => {
    let fundingSourceValues: IProjectFundingFormArrayItem;

    if (itemIndex < funding.fundingSources.length) {
      // edit an existing funding source
      const fundingSource = funding.fundingSources[itemIndex];

      fundingSourceValues = {
        id: fundingSource.id,
        agency_id: fundingSource.agency_id,
        investment_action_category: fundingSource.investment_action_category,
        investment_action_category_name: fundingSource.investment_action_category_name,
        agency_project_id: fundingSource.agency_project_id,
        funding_amount: fundingSource.funding_amount,
        start_date: getFormattedDate(DATE_FORMAT.ShortDateFormat, fundingSource.start_date),
        end_date: getFormattedDate(DATE_FORMAT.ShortDateFormat, fundingSource.end_date),
        revision_count: fundingSource.revision_count
      };
    } else {
      // add a new funding source
      fundingSourceValues = ProjectFundingFormArrayItemInitialValues;
    }

    setFundingFormData({ index: itemIndex, values: fundingSourceValues });

    setOpenEditDialog(true);
  };

  const handleDialogEditSave = async (values: IProjectFundingFormArrayItem) => {
    const projectData = {
      funding: {
        fundingSources: [{ ...values }]
      }
    };

    const isEditing = fundingFormData.index < funding.fundingSources.length;
    const errorTitle = isEditing ? EditFundingI18N.editErrorTitle : AddFundingI18N.addErrorTitle;

    try {
      if (isEditing) {
        await restorationTrackerApi.project.updateProject(id, projectData);
      } else {
        await restorationTrackerApi.project.addFundingSource(id, projectData.funding.fundingSources[0]);
      }

      setOpenEditDialog(false);

      props.refresh();
    } catch (error) {
      const apiError = error as APIError;

      showErrorDialog({ dialogTitle: errorTitle, dialogText: apiError.message, open: true });
    }
  };

  const handleDeleteDialogOpen = async (itemIndex: number) => {
    setFundingFormData({
      index: itemIndex,
      values: funding.fundingSources[fundingFormData.index]
    });
    showYesNoDialog({ open: true });
  };

  const handleDeleteDialogYes = async () => {
    const fundingSource = funding.fundingSources[fundingFormData.index];

    try {
      await restorationTrackerApi.project.deleteFundingSource(id, fundingSource.id);
      showYesNoDialog({ open: false });
    } catch (error) {
      const apiError = error as APIError;
      showErrorDialog({
        dialogTitle: DeleteProjectFundingI18N.deleteErrorTitle,
        dialogText: apiError.message,
        open: true
      });
      return;
    }

    props.refresh();
  };

  const hasFundingSources = funding.fundingSources && funding.fundingSources.length > 0;

  return (
    <>
      <EditDialog
        dialogTitle={
          fundingFormData.index < funding.fundingSources.length ? EditFundingI18N.editTitle : AddFundingI18N.addTitle
        }
        open={openEditDialog}
        component={{
          element: (
            <ProjectFundingItemForm
              funding_sources={
                codes?.funding_source?.map((item) => {
                  return { value: item.id, label: item.name };
                }) || []
              }
              investment_action_category={
                codes?.investment_action_category?.map((item) => {
                  return { value: item.id, fs_id: item.fs_id, label: item.name };
                }) || []
              }
            />
          ),
          initialValues: fundingFormData.values,
          validationSchema: ProjectFundingFormArrayItemYupSchema
        }}
        onCancel={() => setOpenEditDialog(false)}
        onSave={handleDialogEditSave}
      />

      <H3ButtonToolbar
        label="Funding Sources"
        buttonLabel="Add Funding Source"
        buttonTitle="Add Funding Source"
        buttonStartIcon={<Icon path={mdiPlus} size={0.875} />}
        buttonOnClick={() => handleDialogEditOpen(funding.fundingSources.length)}
        toolbarProps={{ disableGutters: true }}
      />

      <TableContainer>
        <Table padding="default" className={classes.fundingSourceTable}>
          <TableHead>
            <TableRow>
              <TableCell>Agency</TableCell>
              <TableCell width="100px">Project ID</TableCell>
              <TableCell>Amount</TableCell>
              <TableCell>Dates</TableCell>
              <TableCell width="130px" align="center">
                Actions
              </TableCell>
            </TableRow>
          </TableHead>

          <TableBody>
            {hasFundingSources &&
              funding.fundingSources.map((item: any, index: number) => (
                <TableRow key={item.id}>
                  <TableCell>
                    {item.agency_name}
                    {item.investment_action_category_name !== 'Not Applicable' && (
                      <Typography component="em" variant="body2">
                        &nbsp;({item.investment_action_category_name})
                      </Typography>
                    )}
                  </TableCell>
                  <TableCell>
                    <Typography variant="body2" noWrap>
                      {item.agency_project_id || 'No Agency Project ID'}
                    </Typography>
                  </TableCell>
                  <TableCell>{getFormattedAmount(item.funding_amount)}</TableCell>
                  <TableCell>
                    <Typography variant="body2" noWrap>
                      {getFormattedDateRangeString(DATE_FORMAT.ShortMediumDateFormat, item.start_date, item.end_date)}
                    </Typography>
                  </TableCell>
                  <TableCell align="center">
                    <IconButton
                      onClick={() => handleDialogEditOpen(index)}
                      title="Edit Funding Source"
                      aria-label="Edit Funding Source"
                      data-testid="edit-funding-source">
                      <Icon path={mdiPencilOutline} size={0.875} />
                    </IconButton>
                    <IconButton
                      data-testid="delete-funding-source"
                      onClick={() => handleDeleteDialogOpen(index)}
                      title="Remove Funding Source"
                      aria-label="Remove Funding Source">
                      <Icon path={mdiTrashCanOutline} size={0.875} />
                    </IconButton>
                  </TableCell>
                </TableRow>
              ))}

            {!hasFundingSources && (
              <TableRow>
                <TableCell colSpan={5}>No Funding Sources</TableCell>
              </TableRow>
            )}
          </TableBody>
        </Table>
      </TableContainer>
    </>
  );
};

export default FundingSource;
