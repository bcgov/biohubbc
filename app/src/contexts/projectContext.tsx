import { useBiohubApi } from 'hooks/useBioHubApi';
import useDataLoader, { DataLoader } from 'hooks/useDataLoader';
import { IGetProjectAttachmentsResponse, IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import { IGetSurveyListResponse } from 'interfaces/useSurveyApi.interface';
import { createContext, PropsWithChildren, useEffect, useMemo } from 'react';
import { useParams } from 'react-router';

/**
 * Context object that stores information about a project
 *
 * @export
 * @interface IProjectContext
 */
export interface IProjectContext {
  /**
   * The Data Loader used to load project data
   *
   * @type {DataLoader<[project_id: number], IGetProjectForViewResponse, unknown>}
   * @memberof IProjectContext
   */
  projectDataLoader: DataLoader<[project_id: number], IGetProjectForViewResponse, unknown>;

  /**
   * The Data Loader used to load project data
   *
   * @type {DataLoader<[project_id: number], IGetSurveyListResponse, unknown>}
   * @memberof IProjectContext
   */
  surveysListDataLoader: DataLoader<[project_id: number], IGetSurveyListResponse, unknown>;

  /**
   * The Data Loader used to load project data
   *
   * @type {DataLoader<[project_id: number], IGetProjectAttachmentsResponse, unknown>}
   * @memberof IProjectContext
   */
  artifactDataLoader: DataLoader<[project_id: number], IGetProjectAttachmentsResponse, unknown>;

  /**
   * The ID belonging to the current project
   *
   * @type {number}
   * @memberof IProjectContext
   */
  projectId: number;
}

export const ProjectContext = createContext<IProjectContext>({
  projectDataLoader: {} as DataLoader<[project_id: number], IGetProjectForViewResponse, unknown>,
  surveysListDataLoader: {} as DataLoader<[project_id: number], IGetSurveyListResponse, unknown>,
  artifactDataLoader: {} as DataLoader<[project_id: number], IGetProjectAttachmentsResponse, unknown>,
  projectId: -1
});

export const ProjectContextProvider = (props: PropsWithChildren<Record<never, any>>) => {
  const biohubApi = useBiohubApi();
  const projectDataLoader = useDataLoader(biohubApi.project.getProjectForView);
  const surveysListDataLoader = useDataLoader(biohubApi.survey.getSurveysBasicFieldsByProjectId);
  const artifactDataLoader = useDataLoader(biohubApi.project.getProjectAttachments);
  const urlParams: Record<string, string | number | undefined> = useParams();

  if (!urlParams['id']) {
    throw new Error(
      "The project ID found in ProjectContextProvider was invalid. Does your current React route provide an 'id' parameter?"
    );
  }

  const projectId = Number(urlParams['id']);

  /**
   * Refreshes the current project object whenever the current project ID changes
   */
  useEffect(() => {
    if (projectId) {
      projectDataLoader.refresh(projectId);
      surveysListDataLoader.refresh(projectId);
      artifactDataLoader.refresh(projectId);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [projectId]);

  const projectContext: IProjectContext = useMemo(() => {
    return {
      projectDataLoader,
      surveysListDataLoader,
      artifactDataLoader,
      projectId
    };
  }, [projectDataLoader, surveysListDataLoader, artifactDataLoader, projectId]);

  return <ProjectContext.Provider value={projectContext}>{props.children}</ProjectContext.Provider>;
};
