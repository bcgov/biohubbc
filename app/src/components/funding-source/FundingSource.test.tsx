import { cleanup, render } from '@testing-library/react';
import { IProjectContext, ProjectContext } from 'contexts/projectContext';
import { useBiohubApi } from 'hooks/useBioHubApi';
import { DataLoader } from 'hooks/useDataLoader';
import React from 'react';
import { getProjectForViewResponse } from 'test-helpers/project-helpers';
import FundingSource, { IFundingSource } from './FundingSource';
jest.mock('../../hooks/useBioHubApi');

const mockUseBiohubApi = {
  project: {
    updateProject: jest.fn(),
    addFundingSource: jest.fn(),
    deleteFundingSource: jest.fn()
  }
};

const mockBiohubApi = ((useBiohubApi as unknown) as jest.Mock<typeof mockUseBiohubApi>).mockReturnValue(
  mockUseBiohubApi
);

const funding_sources: IFundingSource[] = [
  {
    id: 1,
    start_date: '1900-01-01',
    end_date: '2050-01-01',
    agency_project_id: 'Project ID',
    first_nations_name: 'First Nations'
  },
  {
    id: 2,
    agency_name: 'Agency',
    investment_action_category_name: 'Not Applicable',
    funding_amount: 500,
    start_date: '1900-01-01',
    end_date: '1900-02-02',
    agency_project_id: '1230'
  }
];

describe('FundingSource', () => {
  beforeEach(() => {
    // clear mocks before each test
    mockBiohubApi().project.updateProject.mockClear();
    mockBiohubApi().project.addFundingSource.mockClear();
    mockBiohubApi().project.deleteFundingSource.mockClear();
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const mockProjectContext: IProjectContext = {
      projectDataLoader: { data: getProjectForViewResponse } as DataLoader<any, any, any>,
      artifactDataLoader: { data: null } as DataLoader<any, any, any>,
      surveysListDataLoader: { data: null } as DataLoader<any, any, any>,
      projectId: 1
    };

    const { asFragment } = render(
      <ProjectContext.Provider value={mockProjectContext}>
        <FundingSource funding_sources={funding_sources} />
      </ProjectContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
