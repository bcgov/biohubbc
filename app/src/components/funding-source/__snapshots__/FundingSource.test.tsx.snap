// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundingSource renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-divider"
    >
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <span
            class="MuiTypography-root MuiTypography-body1"
          >
            First Nations
          </span>
        </div>
        <dl
          class="MuiBox-root MuiBox-root-3"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6"
            >
              <dt
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Project ID
              </dt>
              <dd
                class="MuiTypography-root MuiTypography-body1"
              >
                Project ID
              </dd>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6"
            >
              <dt
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Timeline
              </dt>
              <dd
                class="MuiTypography-root MuiTypography-body1"
              >
                Jan 1, 1900 - Jan 1, 2050
              </dd>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
            />
          </div>
        </dl>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-divider"
    >
      <div
        class="MuiBox-root MuiBox-root-4"
      >
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <span
            class="MuiTypography-root MuiTypography-body1"
          >
            Agency
          </span>
        </div>
        <dl
          class="MuiBox-root MuiBox-root-6"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6"
            >
              <dt
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Project ID
              </dt>
              <dd
                class="MuiTypography-root MuiTypography-body1"
              >
                1230
              </dd>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6"
            >
              <dt
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Timeline
              </dt>
              <dd
                class="MuiTypography-root MuiTypography-body1"
              >
                Jan 1, 1900 - Feb 2, 1900
              </dd>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12"
            >
              <dt
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Funding Amount
              </dt>
              <dd
                class="MuiTypography-root MuiTypography-body1"
              >
                $500
              </dd>
            </div>
          </div>
        </dl>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
